File: Demos/ListControlDemoCS/CheckBoxListTestForm.cs
//===============================================================================================================
// System  : EWSoftware Data List Control Demonstration Applications
// File    : CheckBoxListTestForm.cs
// Author  : Eric Woodruff  (Eric@EWoodruff.us)
// Updated : 12/02/2024
// Note    : Copyright 2005-2024, Eric Woodruff, All rights reserved
//
// This is used to demonstrate the CheckBoxList control
//
// This code is published under the Microsoft Public License (Ms-PL).  A copy of the license should be
// distributed with the code and can be found at the project website: https://github.com/EWSoftware/ListControls
// This notice, the author's name, and all copyright notices must remain intact in all applications,
// documentation, and source files.
//
//    Date     Who  Comments
// ==============================================================================================================
// 04/17/2005  EFW  Created the code
//===============================================================================================================

using System.Collections;

namespace ListControlDemoCS
{
    /// <summary>
    /// This is used to demonstrate the CheckBoxList control
    /// </summary>
    internal sealed partial class CheckBoxListTestForm : Form
    {
        #region Private data members
        //=====================================================================

        private readonly List<DemoTable> demoData;
        private readonly List<ProductInfo> productInfo;

        #endregion

        #region Constructor
        //=====================================================================

        /// <summary>
        /// Constructor
        /// </summary>
        public CheckBoxListTestForm()
        {
            InitializeComponent();

            try
            {
                using var dc = new DemoDataContext();

                demoData = [.. dc.DemoTable.OrderBy(d => d.Label)];
                productInfo = [.. dc.ProductInfo.OrderBy(p => p.ProductName)];
            }
            catch(SqlException ex)
            {
                demoData = [];
                productInfo = [];

                MessageBox.Show(ex.Message);
            }

            cboDataSource.SelectedIndex = 0;

            pgProps.SelectedObject = cblDemo;
            pgProps.Refresh();
        }
        #endregion

        #region Event handlers
        //=====================================================================

        /// <summary>
        /// Refresh the display and the checkbox list settings after they have changed
        /// </summary>
        /// <param name="s">The sender of the event</param>
        /// <param name="e">The event arguments</param>
        private void pgProps_PropertyValueChanged(object s, PropertyValueChangedEventArgs e)
        {
            cblDemo.Invalidate();
            cblDemo.Update();
        }

        /// <summary>
        /// Change the data source for the checkbox list
        /// </summary>
        /// <param name="sender">The sender of the event</param>
        /// <param name="e">The event arguments</param>
        private void cboDataSource_SelectedIndexChanged(object sender, EventArgs e)
        {
            // Suspend updates to the checkbox list to speed it up and prevent flickering
            cblDemo.BeginInit();

            // Clear out the prior definitions
            cboColumns.Items.Clear();
            cboColumns.SelectedIndex = -1;

            cblDemo.DataSource = null;
            cblDemo.DisplayMember = cblDemo.ValueMember = String.Empty;
            cblDemo.Items.Clear();

            // We must have the data
            if(demoData.Count == 0 || productInfo.Count == 0)
            {
                MessageBox.Show("Database not found.  It must be located in the demo project's folder.", "Error",
                    MessageBoxButtons.OK, MessageBoxIcon.Stop);

                cblDemo.EndInit();
                return;
            }

            // Data tables, views, and sets are also supported but we won't demonstrate that here
            switch(cboDataSource.SelectedIndex)
            {
                case 0:     // Demo data (List<DemoData>)
                    cblDemo.DisplayMember = nameof(DemoTable.Label);
                    cblDemo.ValueMember = nameof(DemoTable.ListKey);
                    cblDemo.DataSource = demoData;
                    break;

                case 1:     // Product info (List<ProductInfo>)
                    cblDemo.DisplayMember = nameof(ProductInfo.ProductName);
                    cblDemo.ValueMember = nameof(ProductInfo.ProductID);
                    cblDemo.DataSource = productInfo;
                    break;

                case 2:     // Array List
                    ArrayList al = new(100);

                    foreach(var p in productInfo)
                        al.Add(new ListItem(p.ProductID, p.ProductName));

                    cblDemo.DisplayMember = nameof(ListItem.Display);
                    cblDemo.ValueMember = nameof(ListItem.Value);
                    cblDemo.DataSource = al;
                    break;

                case 3:     // Item collection strings
                    // Like the above but we add the strings directly to the checkbox list's Items collection
                    foreach(var p in productInfo)
                        cblDemo.Items.Add(p.ProductName);

                    // The item collection is the data source for this one.  It's a simple string list so there
                    // are no display or value members.
                    break;

                default:
                    break;
            }

            // Resume updates to the checkbox list and display the new set of checkboxes
            cblDemo.EndInit();

            // Load the column names
            if(cblDemo.DataSource != null)
            {
                if(cblDemo.DataSource is ArrayList)
                {
                    cboColumns.Items.Add(nameof(ListItem.Display));
                    cboColumns.Items.Add(nameof(ListItem.Value));
                }
                else
                {
                    Type dataSourceType;

                    if(cboDataSource.SelectedIndex == 0)
                        dataSourceType = typeof(DemoTable);
                    else
                        dataSourceType = typeof(ProductInfo);

                    foreach(var p in dataSourceType.GetProperties())
                        cboColumns.Items.Add(p.Name);
                }
            }

            if(cboColumns.Items.Count == 0)
                cboColumns.Enabled = udcRowNumber.Enabled = btnGetValue.Enabled = false;
            else
                cboColumns.Enabled = udcRowNumber.Enabled = btnGetValue.Enabled = true;
        }

        /// <summary>
        /// Get a value from the checkbox list
        /// </summary>
        /// <param name="sender">The sender of the event</param>
        /// <param name="e">The event arguments</param>
        private void btnGetValue_Click(object sender, EventArgs e)
        {
            // This can be any column from the data source regardless of whether or not it is displayed.  Note
            // that you can also use cblDemo["ColName"] to get a column value from the item indicated by the
            // SelectedIndex property.
            txtValue.Text = $"{cboColumns.Text} = {cblDemo[(int)udcRowNumber.Value, cboColumns.Text]}";
        }

        /// <summary>
        /// Show the current item info when the selected index changes.  For the checkbox list, this happens
        /// whenever a new checkbox item gains the focus.
        /// </summary>
        /// <param name="sender">The sender of the event</param>
        /// <param name="e">The event arguments</param>
        private void cblDemo_SelectedIndexChanged(object sender, EventArgs e)
        {
            // Note that SelectedValue is only valid if there is a data source
            txtValue.Text = $"Index = {cblDemo.SelectedIndex}, Value = {cblDemo.SelectedValue}, Text = {cblDemo.Text}";
        }

        /// <summary>
        /// When the check state of a checkbox in the list changes, this event is raised
        /// </summary>
        /// <param name="sender">The sender of the event</param>
        /// <param name="e">The event arguments</param>
        private void cblDemo_ItemCheckStateChanged(object sender, ItemCheckStateEventArgs e)
        {
            txtValue.Text = $"Index = {e.Index}, Current State = {e.CheckState}";
        }

        /// <summary>
        /// Use or clear the image list
        /// </summary>
        /// <param name="sender">The sender of the event</param>
        /// <param name="e">The event arguments</param>
        private void chkShowImages_CheckedChanged(object sender, EventArgs e)
        {
            if(chkShowImages.Checked)
            {
                // The image list is not usable with themed checkboxes under .NET 1.1
                if(cblDemo.FlatStyle == FlatStyle.System)
                    cblDemo.FlatStyle = FlatStyle.Standard;

                cblDemo.ImageList = ilImages;
            }
            else
                cblDemo.ImageList = null;
        }

        /// <summary>
        /// Get the currently checked item values
        /// </summary>
        /// <param name="sender">The sender of the event</param>
        /// <param name="e">The event arguments</param>
        private void btnCheckedItems_Click(object sender, System.EventArgs e)
        {
            CheckedItemsCollection items = cblDemo.CheckedItems;

            if(items.Count == 0)
                MessageBox.Show("No items are currently checked");
            else
            {
                // The values of individual items can be retrieved using the ValueOf() method on the collection.
                // The check state of the items can be retrieved using the CheckStateOf() method.  The ToString()
                // method returns the values as a comma-separated list.
                MessageBox.Show("The values of the currently checked items are:\r\n" + items.ToString());
            }
        }

        /// <summary>
        /// Get the currently checked item display text values
        /// </summary>
        /// <param name="sender">The sender of the event</param>
        /// <param name="e">The event arguments</param>
        private void btnCheckedItemsText_Click(object sender, EventArgs e)
        {
            CheckedItemsCollection items = cblDemo.CheckedItems;

            if(items.Count == 0)
                MessageBox.Show("No items are currently checked");
            else
            {
                // The display text of individual items can be retrieved using the DisplayTextOf() method on the
                // collection.  The check state of the items can be retrieved using the CheckStateOf() method.
                // The ToDisplayTextString() method returns the display text of the checked items as a
                // comma-separated list.
                MessageBox.Show("The display text of the currently checked items are:\r\n" + items.ToDisplayTextString());
            }
        }

        /// <summary>
        /// Get the indices of the currently checked items.
        /// </summary>
        /// <param name="sender">The sender of the event</param>
        /// <param name="e">The event arguments</param>
        private void btnCheckedIndices_Click(object sender, EventArgs e)
        {
            CheckedIndicesCollection indices = cblDemo.CheckedIndices;

            if(indices.Count == 0)
                MessageBox.Show("No items are currently checked");
            else
                MessageBox.Show("The indices of the currently checked items are:\r\n" + indices.ToString());
        }
        #endregion
    }
}

File: Demos/ListControlDemoVB/CheckBoxListTestForm.vb
'================================================================================================================
' System  : EWSoftware Data List Control Demonstration Applications
' File    : CheckBoxListTestForm.vb
' Author  : Eric Woodruff  (Eric@EWoodruff.us)
' Updated : 12/02/2024
' Note    : Copyright 2005-2024, Eric Woodruff, All rights reserved
'
' This is used to demonstrate the CheckBoxList control
'
' This code is published under the Microsoft Public License (Ms-PL).  A copy of the license should be
' distributed with the code and can be found at the project website: https://github.com/EWSoftware/ListControls
' This notice, the author's name, and all copyright notices must remain intact in all applications,
' documentation, and source files.
'
'    Date     Who  Comments
' ===============================================================================================================
' 10/27/2005  EFW  Created the code
'================================================================================================================

Public Partial Class CheckBoxListTestForm
    Inherits Form

    Private ReadOnly demoData As List(Of DemoTable)
    Private ReadOnly productInfo As List(Of ProductInfo)

    Public Sub New()
        MyBase.New()

        InitializeComponent()

        Try
            Using dc As New DemoDataContext()
                demoData = dc.DemoTable.OrderBy(Function (d) d.Label).ToList()
                productInfo = dc.ProductInfo.OrderBy(Function (p) p.ProductName).ToList()
            End Using

        Catch ex As SqlException
            MessageBox.Show(ex.Message)

        End Try

        cboDataSource.SelectedIndex = 0

        pgProps.SelectedObject = cblDemo
        pgProps.Refresh()

    End Sub

    ' Refresh the display and the checkbox list settings after they have changed
    Private Sub pgProps_PropertyValueChanged(s As Object, e As PropertyValueChangedEventArgs) _
      Handles pgProps.PropertyValueChanged
        cblDemo.Invalidate()
        cblDemo.Update()
    End Sub

    ' Change the data source for the checkbox list
    Private Sub cboDataSource_SelectedIndexChanged(sender As Object, e As EventArgs) _
        Handles cboDataSource.SelectedIndexChanged

        ' Suspend updates to the checkbox list to speed it up and prevent flickering
        cblDemo.BeginInit()

        ' Clear out the prior definitions
        cboColumns.Items.Clear()
        cboColumns.SelectedIndex = -1

        cblDemo.DataSource = Nothing
        cblDemo.DisplayMember = String.Empty
        cblDemo.ValueMember = String.Empty
        cblDemo.Items.Clear()

        ' We must have the data
        If demoData.Count = 0 Or productInfo.Count = 0 Then
            MessageBox.Show("Database not found.  It must be located in the demo project's folder.", "Error",
                MessageBoxButtons.OK, MessageBoxIcon.Stop)

            cblDemo.EndInit()
            Return
        End If

        ' Data tables, views, and sets are also supported but we won't demonstrate that here
        Select Case cboDataSource.SelectedIndex
            Case 0      ' Demo data (List(Of DemoData))
                cblDemo.DisplayMember = nameof(DemoTable.Label)
                cblDemo.ValueMember = nameof(DemoTable.ListKey)
                cblDemo.DataSource = demoData

            Case 1      ' Product info (List(Of ProductInfo))
                cblDemo.DisplayMember = nameof(DataBase.ProductInfo.ProductName)
                cblDemo.ValueMember = nameof(Database.ProductInfo.ProductID)
                cblDemo.DataSource = productInfo

            Case 2      ' Array List
                Dim al As New ArrayList(100)

                For Each p In productInfo
                    al.Add(New ListItem(p.ProductID, p.ProductName))
                Next

                cblDemo.DisplayMember = nameof(ListItem.Display)
                cblDemo.ValueMember = nameof(ListItem.Value)
                cblDemo.DataSource = al

            Case 3      ' Item collection strings
                ' Like the above but we add the strings directly to the radio button list's Items collection
                For Each p In productInfo
                    cblDemo.Items.Add(p.ProductName)
                Next

                ' The item collection is the data source for this one.  It's a simple string list so there are no
                ' display or value members.

        End Select

        ' Resume updates to the radio button list and display the new set of radio buttons
        cblDemo.EndInit()

        ' Load the column names
        If Not (cblDemo.DataSource Is Nothing) Then
            If TypeOf(cblDemo.DataSource) Is ArrayList Then
                cboColumns.Items.Add(nameof(ListItem.Display))
                cboColumns.Items.Add(nameof(ListItem.Value))
            Else
                Dim dataSourceType As Type

                If cboDataSource.SelectedIndex = 0
                    dataSourceType = GetType(DemoTable)
                Else
                    dataSourceType = GetType(ProductInfo)
                End If

                For Each p in dataSourceType.GetProperties()
                    cboColumns.Items.Add(p.Name)
                Next
            End If
        End If

        If cboColumns.Items.Count = 0 Then
            cboColumns.Enabled = False
            udcRowNumber.Enabled = False
            btnGetValue.Enabled = False
        Else
            cboColumns.Enabled = True
            udcRowNumber.Enabled = True
            btnGetValue.Enabled = True
        End If
    End Sub

    ' Get a value from the checkbox list
    Private Sub btnGetValue_Click(sender As Object, e As EventArgs) _
      Handles btnGetValue.Click
        ' This can be any column from the data source regardless of whether or not it is displayed.  Note that
        ' you can also use cblDemo("ColName") to get a column value from the item indicated by the SelectedIndex
        ' property.
        txtValue.Text = $"{cboColumns.Text} = {cblDemo(CType(udcRowNumber.Value, Integer), cboColumns.Text)}"
    End Sub

    ' Show the current item info when the selected index changes.  For the checkbox list, this happens whenever a
    ' new checkbox item gains the focus.
    Private Sub cblDemo_SelectedIndexChanged(sender As Object, e As EventArgs) _
      Handles cblDemo.SelectedIndexChanged
        ' Note that SelectedValue is only valid if there is a data source
        txtValue.Text = $"Index = {cblDemo.SelectedIndex}, Value = {cblDemo.SelectedValue}, Text = {cblDemo.Text}"
    End Sub

    ' When the check state of a checkbox in the list changes, this event is raised
    Private Sub cblDemo_ItemCheckStateChanged(sender As Object, e As EWSoftware.ListControls.ItemCheckStateEventArgs) _
      Handles cblDemo.ItemCheckStateChanged
        txtValue.Text = $"Index = {e.Index}, Current State = {e.CheckState}"
    End Sub

    ' Use or clear the image list
    Private Sub chkShowImages_CheckedChanged(sender As Object, e As EventArgs) _
      Handles chkShowImages.CheckedChanged
        If chkShowImages.Checked = True Then
            ' The image list is not usable with themed checkboxes under .NET 1.1
            If cblDemo.FlatStyle = FlatStyle.System Then
                cblDemo.FlatStyle = FlatStyle.Standard
            End If

            cblDemo.ImageList = ilImages
        Else
            cblDemo.ImageList = Nothing
        End If
    End Sub

    ' Get the currently checked item values
    Private Sub btnCheckedItems_Click(sender As Object, e As EventArgs) _
      Handles btnCheckedItems.Click
        Dim items As CheckedItemsCollection = cblDemo.CheckedItems

        If items.Count = 0 Then
            MessageBox.Show("No items are currently checked")
        Else
            ' The values of individual items can be retrieved using the ValueOf() method on the collection.  The
            ' check state of the items can be retrieved using the CheckStateOf() method.  The ToString() method
            ' returns the values as a comma-separated list.
            MessageBox.Show("The values of the currently checked items are:" & Environment.NewLine & items.ToString())
        End If
    End Sub

    ' Get the currently checked item display text values
    Private Sub btnCheckedItemsText_Click(ByVal sender As System.Object, ByVal e As EventArgs) _
        Handles btnCheckedItemsText.Click

        Dim items As CheckedItemsCollection = cblDemo.CheckedItems

        If items.Count = 0 Then
            MessageBox.Show("No items are currently checked")
        Else
            ' The display text of individual items can be retrieved using the DisplayTextOf() method on the
            ' collection.  The check state of the items can be retrieved using the CheckStateOf() method.  The
            ' ToDisplayTextString() method returns the display text of the checked items as a comma-separated
            ' list.
            MessageBox.Show("The display text of the currently checked items are:" & Environment.NewLine &
                items.ToDisplayTextString())
        End If
    End Sub

    ' Get the indices of the currently checked items
    Private Sub btnCheckedIndices_Click(sender As Object, e As EventArgs) _
      Handles btnCheckedIndices.Click
        Dim indices As CheckedIndicesCollection = cblDemo.CheckedIndices

        If indices.Count = 0 Then
            MessageBox.Show("No items are currently checked")
        Else
            MessageBox.Show("The indices of the currently checked items are:" & Environment.NewLine & indices.ToString())
        End If
    End Sub

End Class
